bin/i18n/resources/dashboard/text_to_speech/sync_out.rb

```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/env ruby
+
+require 'json'
+
+require_relative '../../../../../dashboard/config/environment'
+require_relative '../../../../../dashboard/app/models/concerns/text_to_speech'
+require_relative '../../../i18n_script_utils'
+require_relative '../../../utils/sync_out_base'
+require_relative '../text_to_speech'
+
+module I18n
+  module Resources
+    module Dashboard
+      module TextToSpeech
+        class SyncOut < I18n::Utils::SyncOutBase
+          TTS_LOCALES = (::TextToSpeech::VOICES.keys - %i[en-US]).freeze
+          METRIC_CONTEXT = 'update_level_i18n'.freeze
+
+          def perform
+            progress_bar.start
+
+            progress_bar.total = tts_units.size
+            tts_units.find_each do |unit|
+              next unless unit.text_to_speech_enabled?
+
+              unit.levels.merge(Blockly.all).find_each do |level|
+                I18nScriptUtils.process_in_threads(TTS_LOCALES) do |locale|
+                  upload_tts_short_instructions_l10n(level, locale)
+                  upload_tts_long_instructions_l10n(level, locale) unless unit.csf_international? || unit.twenty_hour?
+                  upload_tts_authored_hints_l10n(level, locale)
+                end
+              end
+            ensure
+              progress_bar.increment
+            end
+
+            progress_bar.finish
+          end
+
+          private
+
+          def tts_units
+            @tts_units ||= Unit.where('properties->"$.tts" IS TRUE')
```

Comment:
I just wanted to optimize the db request for the future when we have more units with tts enabled.
So instead of loading all million (potential) records from the db at once and filter them on ruby side, select and process the needed units in batches by 1000 records

Reply:
I could see that being a valuable optimization, but this means that if at any point in the future we want to change the details of how this feature is implemented, we now have to remember to update not just the helper methods that are defined on the Unit class itself but also this i18n sync script. And I'd say that's a lot more likely to happen than us ever having a million Units in the database!

What do you think about implementing this optimization as an ActiveRecord `scope`? We could then update [the `text_to_speech_enabled?` method](https://github.com/code-dot-org/code-dot-org/blob/2c0066db7cfe18df5c3e4752f45ae04740db5baa/dashboard/app/models/unit.rb#L980-L982) to also example the scope, and then everything would be both performant and consistent!